### PR TITLE
Core/SmartScripts: TARGET_INVOKER not targeting passengers of vehicle

### PR DESCRIPTION
**Pull request by tje3d**
This will fix quest’s like “The Black Knight's Orders”, “A Rough Ride”

**Target branch(es):** 3.3.5

### CODE REVIEW

**Comment by Treeston** [comment]
Why should any invoker target  target all passengers?

**Comment by Aokromes** [comment]
i suppose quest / achievements credits.

**Comment by jackpoz** [comment]
would it make more sense to add a new TARGET_* ?

**Comment by tje3d** [comment]
@Treeston just what Aokromes said. For example in "The Black Knight's Orders" quest, Target of event_type WAYPOINT_REACHED is always null and that's why the quest will not be completed ( Same for "A Rough Ride" ).
This is just a try to find proper way.

**Comment by Aokromes** [comment]
Maybe even this can fix achievement credits on ulduar on flame leviatan.

**Comment by Treeston** [comment]
> would it make more sense to add a new TARGET_* ?
^this. `INVOKER` arbitrarily targeting vehicle passengers is counter-intuitive.

**Comment by Killyana** [comment]
Fix https://github.com/TrinityCore/TrinityCore/issues/20974 probably.

**Review by Treeston — CHANGES_REQUESTED** [review]
As was said: Create a new target type `SMART_TARGET_PASSENGER`, preferably with a bitmask indicating which passengers to target.

**Comment by Aokromes** [comment]
What about the ones on c++? (flame leviathan)

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 1]
```diff
@@ -2438,6 +2438,12 @@ void SmartScript::GetTargets(ObjectVector& targets, SmartScriptHolder const& e,
         case SMART_TARGET_ACTION_INVOKER:
             if (scriptTrigger)
                 targets.push_back(scriptTrigger);
+
+            if (me && IsUnit(me))
```
Comment (Shauren) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: but remove it for INVOKER

**Review by Shauren — CHANGES_REQUESTED** [review]

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 2]
```diff
@@ -2696,6 +2696,15 @@ void SmartScript::GetTargets(ObjectVector& targets, SmartScriptHolder const& e,
             }
             break;
         }
+        case SMART_TARGET_PASSENGER:
+        {
```
Comment (Treeston) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: Why braces?

File: src/server/game/AI/SmartScripts/SmartScriptMgr.h [hunk 3]
```diff
@@ -1188,8 +1188,9 @@ enum SMARTAI_TARGETS
     SMART_TARGET_LOOT_RECIPIENTS                = 27,   // all players that have tagged this creature (for kill credit)
     SMART_TARGET_FARTHEST                       = 28,   // maxDist, playerOnly, isInLos
     SMART_TARGET_VEHICLE_ACCESSORY              = 29,   // seat number (vehicle can target it's own accessory)
+    SMART_TARGET_PASSENGER                      = 30,   // passenger's of vehicle
```
Comment (Treeston) [inline src/server/game/AI/SmartScripts/SmartScriptMgr.h]: passengers

**Review by Treeston — COMMENTED** [review]
would still like a bitmask that lets you pick seats to target

**Comment by tje3d** [comment]
@Treeston I've no idea how to implement bitmask 🤔

**Comment by Treeston** [comment]
`if (!mask || (mask & (1 << seatId)))`

**Comment by tje3d** [comment]
@Treeston is this ok?

**Comment by Shauren** [comment]
@Treeston don't you think this makes SMART_TARGET_VEHICLE_ACCESSORY redundant?

**Review by Shauren — APPROVED** [review]

**Comment by Treeston** [comment]
@Shauren are you assuming I understand smartscript?

_(but yes, probably)_

File: src/server/game/AI/SmartScripts/SmartScriptMgr.h [hunk 4]
```diff
@@ -1317,6 +1318,11 @@ struct SmartTarget
         {
             uint32 seat;
         } vehicle;
+
+        struct
+        {
+            int8 seatId;
```
Comment (Treeston) [inline src/server/game/AI/SmartScripts/SmartScriptMgr.h]: `uint32`

File: src/server/game/AI/SmartScripts/SmartScriptMgr.h [hunk 5]
```diff
@@ -1317,6 +1318,11 @@ struct SmartTarget
         {
             uint32 seat;
         } vehicle;
```
Comment (Treeston) [inline src/server/game/AI/SmartScripts/SmartScriptMgr.h]: what is this union member used for?

Comment (tje3d) [reply src/server/game/AI/SmartScripts/SmartScriptMgr.h]: Used by SMART_TARGET_VEHICLE_ACCESSORY:
`
if (Unit* target = me->GetVehicleKit()->GetPassenger(e.target.vehicle.seat))
`

Comment (Treeston) [reply src/server/game/AI/SmartScripts/SmartScriptMgr.h]: how many scripts in current 3.3.5 tdb use this target?

Comment (tje3d) [reply src/server/game/AI/SmartScripts/SmartScriptMgr.h]: Nothing at all

Comment (Treeston) [reply src/server/game/AI/SmartScripts/SmartScriptMgr.h]: remove it then (replace it in the enum with `_UNUSED`)

**Comment by Killyana** [comment]
SMART_TARGET_VEHICLE_ACCESSORY is exactly the same target type, also "The Black Knight's Orders" and  "A Rough Ride" were working correctly months ago, better to find what change broke this.

**Comment by Shauren** [comment]
ok, taking over for merge

**Comment by tje3d** [comment]
@Killyana but the current SMART_TARGET_PASSENGER can do it better now

In those case, WaypointReached always pass `nullptr` as unit, so the TARGET_TYPE_ACTION_INVOKER would not works ( but TARGET_TYPE_PLAYER_RANGE, TARGET_TYPE_PLAYER_DISTANCE and ... would works )

**Comment by Killyana** [comment]
>In those case, WaypointReached always pass nullptr as unit, so the TARGET_TYPE_ACTION_INVOKER would not works.
And it's normal?

**Comment by Treeston** [comment]
I see no reason why `WaypointReached` should get any invoker passed to it.

**Comment by tje3d** [comment]
I will send pull request's for those quest's after the pull req merged.

**Comment by Killyana** [comment]
If I understand you're trying to fix this issue https://github.com/TrinityCore/TrinityCore/issues/20310  by modifying the existing SAI ?

**Comment by tje3d** [comment]
@Killyana Yes at the first 😁

**Comment by Killyana** [comment]
Bad Idea, the issue must be fixed on core side.

**Comment by Shauren** [comment]
I have no objections to extending this sai target, even if it does't fix the issue

**Comment by tje3d** [comment]
@Killyana but as @Treeston  said, maybe its not good idea to use ACTION_INVOKER with WaypointReached.
If it should works with existing AI, then i think we have to open new PULL.

**Comment by Treeston** [comment]
@Killyana please present your reasoning for why `INVOKER` should arbitrarily select the vehicle's passenger for `WAYPOINT_REACHED`, which by design is not invoked by anyone?

**Review by Treeston — APPROVED** [review]

**Comment by Killyana** [comment]
A rough ride: the player cast 51258 which trigger 51257, it will summon the npc 28308 (it's considered as a guardian) on waypoint reached the vehicle will execute quest explored (action 15) on target 7, but after the last changes on movement generator it doesn't work any more, and it's a common script used by many quests.

**Comment by Treeston** [comment]
How does the owner of the guardian become the invoker of the waypoint reached event? Invoker is only set for interaction events.

**Comment by meji46** [comment]
But there isn't invoker for waypoint reached event.

It's better to use SMART_TARGET_OWNER_OR_SUMMONER or SMART_TARGET_STORED in these cases.

**Comment by Killyana** [comment]
All I know it was working before:
``` sql
SELECT * FROM smart_scripts WHERE entryorguid=28308;

**Comment by Treeston** [comment]
> All I know it was working before:

Don't rely on bugs.

**Comment by tje3d** [comment]
@Treeston  Is it good idea to check EVENT_TYPE in SMART_TARGET_ACTION_INVOKER? so we can check if the event was SMART_EVENT_WAYPOINT_* then we use OWNER_OR_SUMMONER as target? 
And this would fix current SAI ( What @Killyana  wants 😉 )

**Comment by Treeston** [comment]
Could we? Yes.
Should we? Hell no, that's utterly counter-intuitive.

**Pull request merged**